### PR TITLE
chore: defork cloud-provider-azure

### PR DESCRIPTION
**Pull request by comtalyst**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
In the past, Azure provider has borrowed some code from [cloud-provider-azure](https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master), although some did not even exist in cloud-provider-azure at that time. 
This results in several pieces of code being duplicated or branched away. This PR is delegating those functionalities back to cloud-provider-azure as much as possible in attempt to eliminate the duplicated logic.

Examples for instances where we miss cloud-provider-azure updates for the duplicated code:
- The authentication logic using AAD certificate, and not to mention security concerns implied from outdated code in this category.
- The config file format (i.e., azure.json) for autoscaler branches out from the format expected by other components using cloud-provider-azure. For example, the field `VmssCacheTTL` in autoscaler's duplicated code is named `VmssCacheTTLInSeconds` in cloud-provider-azure. This causes confusion and kills the purpose of reusable Azure configuration file that cloud-provider-azure is trying to achieve. 

On the other way around:
- #4295, which, arguably, should have been added to cloud-provider-azure instead if there was no duplication in the first place. Proposing changes there: https://github.com/kubernetes-sigs/cloud-provider-azure/issues/6730

Furthermore, we see no need to have a customized version of cloud-provider-azure library for Cluster Autoscaler. Thus, delegating it back is justified. 

This attempt, however, will not be perfect, given we have yet to finish the deprecation for existing behavior differences (e.g., config file format, CLI creds support).

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

`azure_config.go` and `azure_client.go` are the core changes of this PR. Changes in other files are more of consequences.

#### Does this PR introduce a user-facing change?

```release-note
- Azure: from now on, users should refer to https://cloud-provider-azure.sigs.k8s.io/install/configs/ for configuration interface
- Azure: fixed an issue where environment variables were not being passed in when config file exists
- Azure: fixed an issue where some cloud provider configurations were not being validated when UseManagedIdentityExtension is set to true
- Azure: renamed several fields from config file, with old names are still acceptable and taking precedence: `useWorkloadIdentityExtension` to `useFederatedWorkloadIdentityExtension`, `vmssCacheTTL` to `vmssCacheTTLInSeconds`, `vmssVmsCacheTTL` to `vmssVirtualMachinesCacheTTLInSeconds`, `enableVmssFlex` to `enableVmssFlexNodes`
- Azure: renamed several environment variables, with old names are still acceptable and taking precedence: `ARM_USE_MANAGED_IDENTITY_EXTENSION` to `ARM_USE_FEDERATED_WORKLOAD_IDENTITY_EXTENSION`, `AZURE_VMSS_CACHE_TTL` to `AZURE_VMSS_CACHE_TTL_IN_SECONDS`, `AZURE_VMSS_VMS_CACHE_TTL` to `AZURE_VMSS_VMS_CACHE_TTL_IN_SECONDS`, `AZURE_ENABLE_VMSS_FLEX` to `AZURE_ENABLE_VMSS_FLEX_NODES`
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
```

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go [hunk 1]
```diff
@@ -211,18 +211,24 @@ func (as *AgentPool) TargetSize() (int, error) {
...
+	allDeployments, err := as.manager.azureClients.deploymentClient.List(ctx, as.manager.config.ResourceGroup)
 	if err != nil {
-		klog.Errorf("getAllSucceededAndFailedDeployments: failed to list succeeded or failed deployments with error: %v", err)
-		return nil, err
+		klog.Errorf("getAllSucceededAndFailedDeployments: failed to list deployments with error: %v", err)
+		return nil, err.Error()
+	}
+
+	result := make([]resources.DeploymentExtended, 0)
+	for _, deployment := range allDeployments {
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go:225]: Could have small performance impact from client-side filtering (not sure if the original filter is server-side). 
Should be neglectable unless I am missing something. Plus VMAS pool is not properly supported by AKS anymore (will be discussed separately)

File: cluster-autoscaler/cloudprovider/azure/azure_config.go [hunk 2]
```diff
@@ -81,70 +64,49 @@ type CloudProviderRateLimitConfig struct {
...
 type Config struct {
-	CloudProviderRateLimitConfig
-
-	Cloud          string `json:"cloud" yaml:"cloud"`
-	Location       string `json:"location" yaml:"location"`
-	TenantID       string `json:"tenantId" yaml:"tenantId"`
-	SubscriptionID string `json:"subscriptionId" yaml:"subscriptionId"`
-	ResourceGroup  string `json:"resourceGroup" yaml:"resourceGroup"`
-	VMType         string `json:"vmType" yaml:"vmType"`
+	providerazure.Config `json:",inline" yaml:",inline"`
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_config.go:67]: Delegated most of the config back to `cloud-provider-azure`.

However, this, along with other **user-facing** interfaces that got delegated back to `cloud-provider-azure` will be more hideous from the user. For example, if VM type string `"vmss"` changes to `"vmscalesets"` on the newer version of `cloud-provider-azure`, it won't be visible to the user that might not be aware of the usage of `cloud-provider-azure`. 

Well-covered unit testing would definitely help still. But it might even be difficult for the maintainers to keep track of the true size of the exposed interface by the library (e.g., through this `Config` and `cloud-config-file` which will be unmarshalled into this struct).

Or, maybe it is by design that the user is supposed to know the existence of this module and can configure them? If so, is it intended for the maintainer to not having to worry about those interfaces, and treat them as black boxes?

Anyway, in this case (and likely in most cases--in practice), although not ideal, I think it is worth it to delegate it back given that **changes are really rare**, the size of the interface is not too large to have a proper coverage/know what to care about, and there should be a proper deprecation procedures if something changes.

File: cluster-autoscaler/cloudprovider/azure/azure_config.go [hunk 3]
```diff
@@ -42,27 +43,9 @@ const (
 
 	imdsServerURL = "http://169.254.169.254"
 
-	// backoff
-	backoffRetriesDefault  = 6
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_config.go:46]: Delegated back to `cloud-provider-azure`

File: cluster-autoscaler/cloudprovider/azure/azure_config.go [hunk 4]
```diff
@@ -154,177 +116,126 @@ func BuildAzureConfig(configReader io.Reader) (*Config, error) {
...
+	if _, err = assignBoolFromEnvIfExists(&cfg.UseManagedIdentityExtension, "ARM_USE_MANAGED_IDENTITY_EXTENSION"); err != nil {
+		return nil, err
+	}
+	if _, err = assignBoolFromEnvIfExists(&cfg.UseFederatedWorkloadIdentityExtension, "ARM_USE_WORKLOAD_IDENTITY_EXTENSION"); err != nil {
+		return nil, err
+	}
+	if _, err = assignFromEnvIfExists(&cfg.UserAssignedIdentityID, "ARM_USER_ASSIGNED_IDENTITY_ID"); err != nil {
+		return nil, err
+	}
+	if _, err = assignIntFromEnvIfExists(&cfg.VmssCacheTTLInSeconds, "AZURE_VMSS_CACHE_TTL"); err != nil {
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_config.go:171]: This is one of the fields that have a difference in naming (`cloud-provider-azure` version has "InSeconds"). Although we keep the environment variables the same to not introduce breaking changes. Not that environment variable name needs to be changed either, given the minimal change.

File: cluster-autoscaler/cloudprovider/azure/azure_config.go [hunk 5]
```diff
@@ -498,23 +303,25 @@ func (cfg *Config) validate() error {
...
+		default:
+			return fmt.Errorf("unsupported authorization method: %s", cfg.AuthMethod)
 		}
-	case authMethodCLI:
-		// Nothing to check at the moment.
-	default:
-		return fmt.Errorf("unsupported authorization method: %s", cfg.AuthMethod)
 	}
 
 	if cfg.CloudProviderBackoff && cfg.CloudProviderBackoffRetries == 0 {
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_config.go:327]: This used to be unintentionally(?) neglected when `UseManagedIdentityExtension` is `true`, which should be completely unrelated. Current changes should fix this bug.

File: cluster-autoscaler/cloudprovider/azure/azure_config.go [hunk 6]
```diff
@@ -103,56 +70,54 @@ type Config struct {
...
-	AADClientID                  string `json:"aadClientId" yaml:"aadClientId"`
-	AADClientSecret              string `json:"aadClientSecret" yaml:"aadClientSecret"`
-	AADClientCertPath            string `json:"aadClientCertPath" yaml:"aadClientCertPath"`
-	AADClientCertPassword        string `json:"aadClientCertPassword" yaml:"aadClientCertPassword"`
-	AADFederatedTokenFile        string `json:"aadFederatedTokenFile" yaml:"aadFederatedTokenFile"`
-	UseManagedIdentityExtension  bool   `json:"useManagedIdentityExtension" yaml:"useManagedIdentityExtension"`
-	UseWorkloadIdentityExtension bool   `json:"useWorkloadIdentityExtension" yaml:"useWorkloadIdentityExtension"`
-	UserAssignedIdentityID       string `json:"userAssignedIdentityID" yaml:"userAssignedIdentityID"`
+	// 06/19/2024: This field is awkward, given the existence of UseManagedIdentityExtension and UseFederatedWorkloadIdentityExtension.
+	// Ideally, either it should be deprecated, or reworked to be on the same "dimension" as the two above, if not reworking those two.
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_config.go:74]: This comment presumably refers to `AuthMethod`; maybe move in front of it?

File: cluster-autoscaler/cloudprovider/azure/azure_config.go (hunk 2 above)
Comment (tallaxes) [reply cluster-autoscaler/cloudprovider/azure/azure_config.go:55]: "more hideous" -> "more hidden / less visible / less discoverable"? (And I don't think this is a major issue; can be addressed by documentation.)

Comment (tallaxes) [reply cluster-autoscaler/cloudprovider/azure/azure_config.go:55]: It may also be worth capturing in some way whatever guarantees are intended regarding compatibility of this config with cloud-provider-azure config serialization. Though answer is probably just that this one is a superset?

File: cluster-autoscaler/cloudprovider/azure/azure_client.go [hunk 7]
```diff
@@ -17,216 +17,42 @@ limitations under the License.
...
-			env.ServiceManagementEndpoint)
-	}
-
-	return nil, fmt.Errorf("no credentials provided for AAD application %s", config.AADClientID)
-}
-
 func newAuthorizer(config *Config, env *azure.Environment) (autorest.Authorizer, error) {
 	switch config.AuthMethod {
 	case authMethodCLI:
 		return auth.NewAuthorizerFromCLI()
```
[the remark this answers is not in the document]
Comment (tallaxes) [reply cluster-autoscaler/cloudprovider/azure/azure_client.go:177]: Let's propose supporting it in cloud-provider-azure?

**Review by tallaxes — COMMENTED** [review]
I really like where this is going, don't see any blockers (though would need another round of close review). Would be good to find a way to prove there are no regressions - to what extent do you think the tests cover this?

File: cluster-autoscaler/cloudprovider/azure/azure_config.go (hunk 2 above)
Comment (comtalyst) [reply cluster-autoscaler/cloudprovider/azure/azure_config.go:55]: > "more hideous" -> "more hidden / less visible / less discoverable"?

Yes. And also less aware of changes.

> can be addressed by documentation

How would you imagine the documentation to be like? 
Something like A) "we are using cloud-provider-azure, go see their repo for how to use it" or B) "here are the configs" then 'fork' the docs from cloud-provider-azure?
A) would be easier to be maintained, but additional burden to the user. B) would be like a fork to us.

If you ask me I would prefer A). In practice it wouldn't hurt the experience that much given that it seems to always be one-layer.

**Comment by comtalyst** [comment]
> I really like where this is going, don't see any blockers (though would need another round of close review). Would be good to find a way to prove there are no regressions - to what extent do you think the tests cover this?

For config changes, the existing unit tests could be a prove that the delegations are not breaking interface-wise. More unit tests capturing more corner cases will be added as well. I have confidence in those so far.

For service principal token, I think the E2Es are currently the best bet. Some manual long-running tests might help as well.
Looking at their code, I don't see if they have real differences from ours at the moment. But not that my eyes should be trusted that much.

File: cluster-autoscaler/cloudprovider/azure/azure_config.go (hunk 3 above)
Comment (rakechill) [reply cluster-autoscaler/cloudprovider/azure/azure_config.go:46]: have we confirmed that the cloud-provider-azure defaults are the same as those listed here?

File: cluster-autoscaler/cloudprovider/azure/azure_config.go (hunk 2 above)
Comment (rakechill) [reply cluster-autoscaler/cloudprovider/azure/azure_config.go:55]: A) makes sense to me, since `cloud-provider-azure` should be keeping their docs up to date (doesn't make sense for us to duplicate the work here).

File: cluster-autoscaler/cloudprovider/azure/azure_config.go (hunk 3 above)
Comment (comtalyst) [reply cluster-autoscaler/cloudprovider/azure/azure_config.go:46]: Yes I did. 
Also considering about having unit tests to guarantee them in a long run, but may not be necessary if we rely on user to check cloud-provider-azure directly.

File: cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go [hunk 8]
```diff
@@ -82,10 +82,10 @@ func (as *AgentPool) initialize() error {
 	ctx, cancel := getContextWithCancel()
 	defer cancel()
 
-	template, err := as.manager.azClient.deploymentsClient.ExportTemplate(ctx, as.manager.config.ResourceGroup, as.manager.config.Deployment)
+	template, err := as.manager.azClient.deploymentClient.ExportTemplate(ctx, as.manager.config.ResourceGroup, as.manager.config.Deployment)
 	if err != nil {
-		klog.Errorf("deploymentsClient.ExportTemplate(%s, %s) failed: %v", as.manager.config.ResourceGroup, as.manager.config.Deployment, err)
-		return err
+		klog.Errorf("deploymentClient.ExportTemplate(%s, %s) failed: %v", as.manager.config.ResourceGroup, as.manager.config.Deployment, err)
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go:87]: Is passing retry.Error to log still Ok here (and elsewhere)?
(Also, naming it something like `rerr` would make it easier to follow)

File: cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go [hunk 9]
```diff
@@ -317,22 +326,20 @@ func (as *AgentPool) IncreaseSize(delta int) error {
...
-
-		// Update cache after scale success.
-		as.curSize = int64(expectedSize)
-		as.lastRefresh = time.Now()
-		klog.V(6).Info("IncreaseSize: invalidating cache")
-		as.manager.invalidateCache()
-		return nil
+	klog.V(3).Infof("Waiting for deploymentClient.CreateOrUpdate(%s, %s, %v)", as.manager.config.ResourceGroup, newDeploymentName, newDeployment)
+	retryErr := as.manager.azClient.deploymentClient.CreateOrUpdate(ctx, as.manager.config.ResourceGroup, newDeploymentName, newDeployment, "")
+	if err != nil {
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go:331]: Oops (will never happen)

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go [hunk 10]
```diff
@@ -55,18 +57,19 @@ func newTestAzureManager(t *testing.T) *AzureManager {
...
-			VMType:              vmTypeVMSS,
+			Config: providerazure.Config{
+				ResourceGroup: "rg",
+				VMType:        providerazureconsts.VMTypeVMSS,
+			},
 			MaxDeploymentsCount: 2,
 			Deployment:          "deployment",
-			EnableForceDelete:   true,
-			Location:            "eastus",
+			EnableForceDelete:   false,
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go:66]: Why change?

File: cluster-autoscaler/cloudprovider/azure/azure_config.go [hunk 11]
```diff
@@ -532,23 +379,29 @@ func (cfg *Config) validate() error {
...
-		return nil
+	if cfg.UseManagedIdentityExtension && cfg.UseFederatedWorkloadIdentityExtension {
+		return fmt.Errorf("you can not combine both managed identity and workload identity as an authentication mechanism")
 	}
 
-	if cfg.TenantID == "" {
-		return fmt.Errorf("tenant ID not set")
+	if cfg.VMType != providerazureconsts.VMTypeStandard && cfg.VMType != providerazureconsts.VMTypeVMSS {
+		return fmt.Errorf("unsupported VM type: %s", cfg.VMType)
 	}
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_config.go:388]: The comment on VMType still says vmssflex value is supported. Also need to think through migration, if that's indeed how it was intended to be set, and is changing now

File: cluster-autoscaler/cloudprovider/azure/azure_config_test.go
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_config_test.go:1]: Any more tests we could/should add here (e.g. to test the changes in the overrid logic, such as legacy overrides?) Or are these covered in the azure_manager_tests?

File: cluster-autoscaler/cloudprovider/azure/azure_fakes.go
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_fakes.go:1]: There is no suitable mock in the cloud provider?

File: cluster-autoscaler/cloudprovider/azure/azure_scale_set_test.go [hunk 12]
```diff
@@ -549,48 +541,18 @@ func TestDeleteNodes(t *testing.T) {
...
-			name:              "flexible, force delete disabled",
-			orchestrationMode: compute.Flexible,
-			enableForceDelete: false,
-		},
-	}
-
-	for _, tc := range cases {
-		orchMode := tc.orchestrationMode
-		enableForceDelete := tc.enableForceDelete
+	orchestrationModes := [2]compute.OrchestrationMode{compute.Uniform, compute.Flexible}
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_scale_set_test.go:544]: What's the reasoning for removing tests for forcedelete?

File: cluster-autoscaler/cloudprovider/azure/azure_util.go [hunk 13]
```diff
@@ -636,3 +636,21 @@ func vmPowerStateFromStatuses(statuses []compute.InstanceViewStatus) string {
...
+func parseInt32(s string, base int) (int, error) {
+	val, err := strconv.ParseInt(s, base, 32)
+	if err != nil {
+		return 0, err
+	}
+	return int(val), nil
+}
+
+// strconv.ParseFloat, but for float32
+func parseFloat32(s string, base float32) (float32, error) {
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_util.go:650]: base unused

**Review by tallaxes — CHANGES_REQUESTED** [review]
Looks great, mostly minor quips/questions

**Comment by comtalyst** [comment]
> May also need to check if updates/renames to config/environment variables warrant any updates to readme and/or charts (could be a follow-up).

As discussed, I believe we should delegate all those to cloud-provider-azure docs. We could sweep through the readmes if there is any left to begin with. 
For now, I left a note in the release notes (in PR description).

File: cluster-autoscaler/cloudprovider/azure/azure_config.go (hunk 11 above)
Comment (comtalyst) [reply cluster-autoscaler/cloudprovider/azure/azure_config.go:388]:
```go
func (m *AzureManager) buildNodeGroupFromSpec(spec string) (cloudprovider.NodeGroup, error) {
	scaleToZeroSupported := scaleToZeroSupportedStandard
	if strings.EqualFold(m.config.VMType, providerazureconsts.VMTypeVMSS) {
		scaleToZeroSupported = scaleToZeroSupportedVMSS
	}
	s, err := dynamic.SpecFromString(spec, scaleToZeroSupported)
	if err != nil {
		return nil, fmt.Errorf("failed to parse node group spec: %v", err)
	}
	vmsPoolSet := m.azureCache.getVMsPoolSet()
	if _, ok := vmsPoolSet[s.Name]; ok {
		return NewVMsPool(s, m), nil
	}

	switch m.config.VMType {
	case providerazureconsts.VMTypeStandard:
		return NewAgentPool(s, m)
	case providerazureconsts.VMTypeVMSS:
		return NewScaleSet(s, m, -1, false)
	default:
		return nil, fmt.Errorf("vmtype %s not supported", m.config.VMType)
	}
}
```
From this (that has always been here), VM types other than these two will effectively disable CAS already, given this method is critical. I don't think anyone can use that right now.
I think it is fair to "notify users" through fail initialization like this, given the config is cluster-level.

For our support of flex, I think we do have it (that `EnableVMSSFlexNodes`?), but not through this. This sounds like something to revisit.

File: cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go (hunk 8 above)
Comment (comtalyst) [reply cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go:87]: Great advice. Didn't see those coming.

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go (hunk 10 above)
Comment (comtalyst) [reply cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go:66]: A mistake in merge conflict resolving it seems. Fixed along others.

File: cluster-autoscaler/cloudprovider/azure/azure_config_test.go
Comment (comtalyst) [reply cluster-autoscaler/cloudprovider/azure/azure_config_test.go:1]: Those are already covered in azure_manager_test.go, I think the coverage is quite decent now.

File: cluster-autoscaler/cloudprovider/azure/azure_fakes.go
Comment (comtalyst) [reply cluster-autoscaler/cloudprovider/azure/azure_fakes.go:1]: Haven't look at that yet. But, given low traffic for this area (deprecated VMAS type pool), I don't think we should put effort on it now.

**Review by tallaxes — CHANGES_REQUESTED** [review]
LGTM but for error checking bug

File: cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go [hunk 14]
```diff
@@ -258,9 +267,9 @@ func (as *AgentPool) deleteOutdatedDeployments() (err error) {
 	errList := make([]error, 0)
 	for _, deployment := range toBeDeleted {
 		klog.V(4).Infof("deleteOutdatedDeployments: starts deleting outdated deployment (%s)", *deployment.Name)
-		_, err := as.manager.azClient.deploymentsClient.Delete(ctx, as.manager.config.ResourceGroup, *deployment.Name)
+		rerr := as.manager.azClient.deploymentClient.Delete(ctx, as.manager.config.ResourceGroup, *deployment.Name)
 		if err != nil {
```
[the remark this answers is not in the document]
Comment (comtalyst) [reply cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go]: Done, and looks like this one of the last of this kind of issue

**Review by tallaxes — APPROVED** [review]
/lgtm
/approved

**Comment by comtalyst** [comment]
/assign towca
@towca there are some module bumps in go.mod from this, thanks

**Comment by comtalyst** [comment]
/assign @towca
There are some module bumps in go.mod from this, thanks.

**Comment by towca** [comment]
CA go.mod changes LGTM

/approve
